accepted/2020/platform-checks/platform-checks.md

```diff
@@ -0,0 +1,450 @@
+  Version][os-minimum-version] for more details.
+
+.NET has always taken the position that providing platform-specific
+functionality isn't a bug but a feature because it empowers developers to build
+applications that feel native for the user. That's why .NET has rich interop
+facilities such as P/Invoke.
+
+At the same time, developers really dislike the notion of "write once and test
+everywhere" because it results in a delayed feedback loop.
+
+Fortunately .NET has also taken the position that developer productivity is key.
+This document proposes:
+
+1. A mechanism to annotate APIs as being platform-specific, including the
+   version they got introduced in, the version they got removed in, and the
+   version they got obsoleted in.
+
+2. A Roslyn analyzer that informs developers when they use platform-specific
+   APIs from call sites where the API might not be available.
+
+This work draws heavily form the experience of the [API Analyzer] and improves
+on it by making it a first-class platform feature.
+
+## Scenarios and User Experience
+
+### Lighting up on later versions of the OS
+
+Miguel is building Baby Shark, a popular iOS application. He started with .NET
+that supported iOS 13 but Apple just released iOS 14, which adds the great
+`NSFizBuzz` API that gives his app the extra pop.
+
+After upgrading the app and writing the code, Miguel decides that while the
+feature is cool, it doesn't warrant cutting of all his customers who are
+currently still on iOS 13. So he edits the project file and sets
+`TargetPlatformMinVersion` to 13.
+
+```xml
+<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>net5.0-ios14.0</TargetFramework>
+    <TargetPlatformMinVersion>13.0</TargetPlatformMinVersion>
+  </PropertyGroup>
+
+  ...
+
+</Project>
+```
+
+After doing this, he gets a diagnostic in this code:
+
+> 'NSFizzBuff' requires iOS 14 or later.
+
+```C#
+static void ProvideExtraPop()
+{
+    NSFizzBuff();
+}
+```
+
+He invokes the Roslyn code fixer which suggests to guard the call. This results
+in the following code:
+
+```C#
+static void ProvideExtraPop()
+{
+    if (RuntimeInformation.IsOSPlatformOrLater(OSPlatform.iOS, 14))
+        NSFizzBuff();
+}
+```
+
+### Detecting obsoletion of OS APIs
+
+After upgrading to iOS 14 Miguel also got the following diagnostic:
+
+> 'UIApplicationExitsOnSuspend' has been deprecated since iOS 13.1
+
+Miguel decides that he'll tackle that problem later as he needs to ship now.
+
+### Detecting removal of OS APIs
+
+A few year after his massively successful fizz-buzzed Baby Shark, Apple releases
+iOS 15. After upgrading Miguel gets the following diagnostic:
+
+> 'UIApplicationExitsOnSuspend' has been removed since iOS 15.0
+
+Doh! After reading more details, he releases that he can just stop calling this
+API because Apple automatically suspends apps leaving the foreground when they
+don't require background execution. However, since he still supports iOS 12, he
+need
```

Comment:
You trailed off a bit there :)

Reply:
I think it is basically how do you support both the old and the new one. I think also a pragma to say I know what I am doing.